R/expr-expr.R

```diff
@@ -1070,102 +1070,117 @@ expr__filter <- function(...) {
-#' is slower than native polars.
-#' The user function must take one polars `Series` as input and the return
-#' should be a `Series` or any Robj convertible into a `Series` (e.g. vectors).
-#' Map fully supports `browser()`.
-#'
-#' If `in_background = FALSE` the function can access any global variable of the
-#' R session. However, note that several calls to [`$map_batches()`][expr__map_batches]
-#' will sequentially share the same main R session,
-#' so the global environment might change between the start of the query and the moment
-#' a [`$map_batches()`][expr__map_batches] call is evaluated. Any native
-#' polars computations can still be executed meanwhile. If `in_background = TRUE`,
-#' the map will run in one or more other R sessions and will not have access
-#' to global variables. Use `options(polars.rpool_cap = 4)` and
-#' `polars_options()$rpool_cap` to set and view number of parallel R sessions.
-#'
-#' @examples
-#' pl$DataFrame(iris)$
-#'   select(
-#'   pl$col("Sepal.Length")$map_batches(\(x) {
-#'     paste("cheese", as.character(x$to_vector()))
-#'   }, pl$String)
-#' )
-#'
-#' # R parallel process example, use Sys.sleep() to imitate some CPU expensive
-#' # computation.
-#'
-#' # map a,b,c,d sequentially
-#' pl$LazyFrame(a = 1, b = 2, c = 3, d = 4)$select(
-#'   pl$all()$map_batches(\(s) {
-#'     Sys.sleep(.1)
-#'     s * 2
-#'   })
-#' )$collect() |> system.time()
-#'
-#' # map in parallel 1: Overhead to start up extra R processes / sessions
-#' options(polars.rpool_cap = 0) # drop any previous processes, just to show start-up overhead
-#' options(polars.rpool_cap = 4) # set back to 4, the default
-#' polars_options()$rpool_cap
-#' pl$LazyFrame(a = 1, b = 2, c = 3, d = 4)$select(
-#'   pl$all()$map_batches(\(s) {
-#'     Sys.sleep(.1)
-#'     s * 2
-#'   }, in_background = TRUE)
-#' )$collect() |> system.time()
-#'
-#' # map in parallel 2: Reuse R processes in "polars global_rpool".
-#' polars_options()$rpool_cap
-#' pl$LazyFrame(a = 1, b = 2, c = 3, d = 4)$select(
-#'   pl$all()$map_batches(\(s) {
-#'     Sys.sleep(.1)
-#'     s * 2
-#'   }, in_background = TRUE)
-#' )$collect() |> system.time()
-# TODO: remove the noRd tag
-#' @noRd
+#' @examples
+#' df <- pl$DataFrame(
+#'   sine = c(0.0, 1.0, 0.0, -1.0),
+#'   cosine = c(1.0, 0.0, -1.0, 0.0)
+#' )
+#' df$select(pl$all()$map_batches(\(x) {
+#'   elems <- as.vector(x)
+#'   which.max(elems)
+#' }))
+#'
+#' # In a group-by context, the `agg_list` parameter can improve performance if
+#' # used correctly. The following example has `agg_list = FALSE`, which causes
+#' # the function to be applied once per group. The input of the function is a
+#' # Series of type Int64. This is less efficient.
+#' df <- pl$DataFrame(
+#'   a = c(0, 1, 0, 1),
+#'   b = c(1, 2, 3, 4)
+#' )
+#' system.time({
+#'   print(
+#'     df$group_by("a")$agg(
+#'       pl$col("b")$map_batches(\(x) x + 2, agg_list = FALSE)
+#'     )
+#'   )
+#' })
```

Comment:
This is actually faster than the alternative below so far, it might change with the rework on `.DollarNames`